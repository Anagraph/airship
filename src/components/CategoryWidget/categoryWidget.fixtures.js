export default {
  max: 359029623,
  categories: [
    {
      group: false,
      name: 'China',
      value: 359029623,
    },
    {
      group: false,
      name: 'United States of America',
      value: 236513856,
    },
    {
      group: false,
      name: 'India',
      value: 204338075,
    },
    {
      group: false,
      name: 'Brazil',
      value: 128199480,
    },
    {
      group: false,
      name: 'Japan',
      value: 89712598,
    },
    {
      group: false,
      name: 'Russia',
      value: 80644642,
    },
    {
      group: false,
      name: 'Mexico',
      value: 67386841,
    },
    {
      group: false,
      name: 'Indonesia',
      value: 48433879,
    },
    {
      group: false,
      name: 'Pakistan',
      value: 41364699,
    },
    {
      group: true,
      name: 'Other',
      value: 1107313867,
    },
  ],
};
