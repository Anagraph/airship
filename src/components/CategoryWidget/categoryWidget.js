import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { darken, lighten } from 'polished';
import { colors } from '../../constants';
import { readableNumber } from '../../utils';
import Base from '../Typography/base';

const CATEGORY_OTHER = 'Other';

const Categories = styled.ul`
  padding: 0;
  list-style: none;
`;

const Amount = Base.withComponent('span').extend`
  flex: 30%;
  text-align: right;
  font-size: 12px;
  line-height: 20px;
`;

const Name = Base.withComponent('p').extend`
  flex: 70%;
  font-size: 12px;
  line-height: 20px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: left;
`;

const Progress = styled.div`
  height: 4px;
  width: 100%;
  border-radius: 2px;
  background: ${colors.ui02};
  position: relative;

  &::after {
    content: '';
    width: ${props => props.amount}%;
    height: 4px;
    border-radius: 2px;
    transition: background 0.2s ease;
    position: absolute;
    max-width: 100%;
    left: 0;
  }
`;

const Category = styled.li`
  display: flex;
  flex-wrap: wrap;
  cursor: ${props => props.onCategoryClick ? 'pointer' : 'default'};
  margin-bottom: 8px;

  & > ${Progress}::after {
    background: ${props => (props.isOther ? colors.type01 : colors.brand03)};
  }

  ${props => props.onCategoryClick && css`
    &:hover {
      & > ${Progress}::after {
        background: ${props => (props.isOther ? lighten(0.4, colors.type01) : darken(0.16, colors.brand03))};
      }
    }
  `};

  ${props => !props.selected && css`
    ${Amount},
    ${Name} {
      color: ${colors.type02};
    }

    ${Progress}::after {
      background: ${colors.ui03};
    }

    ${props => props.onCategoryClick && `
      &:hover {
        & > ${Progress}::after {
          background: ${colors.ui04}
        }
      }
    `};
  `}
`;

Category.displayName = 'Category';

class CategoryWidget extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selected: props.selected || [],
    };
  }

  onCategoryClick = clickedCategory => {
    if (!this.props.onCategoryClick) return;

    const { selected } = this.state;

    const nextSelected = selected.includes(clickedCategory)
      ? selected.filter(category => category !== clickedCategory)
      : [...selected, clickedCategory];

    this.setState({ selected: nextSelected });

    this.props.onCategoryClick(nextSelected);
  }

  componentWillReceiveProps(nextProps) {
    this.setState(prevState => ({ selected: nextProps.selected || prevState.selected }));
  }

  render() {
    const { categories, max } = this.props.data;
    const { selected } = this.state;

    return (
      <Categories>
        {categories && categories.map(category => {
          const { name, value } = category;
          const isSelected = selected.length === 0 || selected.includes(category);

          return (
            <Category
              key={name}
              onClick={() => this.onCategoryClick(category)}
              selected={isSelected}
              isOther={name === CATEGORY_OTHER}
            >
              <Name>{name}</Name>
              <Amount>{readableNumber(value)}</Amount>
              <Progress amount={(value / max) * 100} />
            </Category>
          );
        })}
      </Categories>
    );
  }
}

CategoryWidget.propTypes = {
  data: PropTypes.shape({
    count: PropTypes.number,
    max: PropTypes.number,
    min: PropTypes.number,
    nulls: PropTypes.number,
    operation: PropTypes.string,
    categories: PropTypes.arrayOf(PropTypes.object),
  }),
  onCategoryClick: PropTypes.func,
  selected: PropTypes.array,
};

CategoryWidget.defaultProps = {
  data: {},
};

export default CategoryWidget;
