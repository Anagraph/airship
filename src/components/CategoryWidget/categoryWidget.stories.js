import React from 'react';
import { storiesOf } from '@storybook/react';
import CategoryWidget from './categoryWidget';
import Widget from '../Widget/widget';
import mockData from './categoryWidget.fixtures';

storiesOf('Category Widget', module)
  .add('Default', () => (
    <CategoryWidget
      categories={mockData.categories}
      max={mockData.max}
    />
  ))
  .add('Inside a widget', () => (
    <Widget>
      <Widget.Title>Widget</Widget.Title>
      <Widget.Description>Just a widget</Widget.Description>

      <CategoryWidget
        categories={mockData.categories}
        max={mockData.max}
      />
    </Widget>
  ))
  .add('Custom color', () => (
    <Widget>
      <Widget.Title>Widget</Widget.Title>
      <Widget.Description>Just a widget</Widget.Description>

      <CategoryWidget
        categories={mockData.categories}
        max={mockData.max}
        color="#FABADA"
      />
    </Widget>
  ));
