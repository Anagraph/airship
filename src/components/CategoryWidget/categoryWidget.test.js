import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import CategoryWidget from './categoryWidget';
import mockData from './categoryWidget.fixtures';

describe('<CategoryWidget />', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<CategoryWidget />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with data', () => {
    const component = renderer.create(<CategoryWidget categories={mockData.categories} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('overrides the category color', () => {
    const component = renderer.create(
      <CategoryWidget
        categories={mockData.categories}
        color="#FABADA"
      />
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('when a category is clicked', () => {
    describe('when onCategoryClick is provided', () => {
      it('adds the category to the state', () => {
        const component = mount(<CategoryWidget categories={mockData.categories} onCategoryClick={() => {}} />);
        const instance = component.instance();

        expect(instance.state.selected.length).toEqual(0);

        component.find('Category').at(1).simulate('click');
        component.update();

        expect(instance.state.selected.length).toEqual(1);
        expect(instance.state.selected[0]).toEqual(mockData.categories[1].name);
      });

      it('calls onCategoryClick with the selected categories', () => {
        const clickMock = jest.fn();
        const component = mount(<CategoryWidget categories={mockData.categories} onCategoryClick={clickMock} />);


        component.find('Category').at(0).simulate('click');
        component.find('Category').at(1).simulate('click');

        expect(clickMock).toHaveBeenCalledWith([
          mockData.categories[0].name,
          mockData.categories[1].name,
        ]);
      });
    });
  });

  describe('when selected prop is changed', () => {
    it('updates the state with the new data', () => {
      const component = mount(
        <CategoryWidget
          categories={mockData.categories}
          selected={mockData.categories}
        />
      );
      const instance = component.instance();

      expect(instance.state.selected.length).toEqual(mockData.categories.length);

      component.setProps({ selected: [] });
      component.update();

      expect(instance.state.selected.length).toEqual(0);
    });
  });
});
