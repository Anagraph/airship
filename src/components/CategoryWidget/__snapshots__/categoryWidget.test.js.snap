// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryWidget /> overrides the category color 1`] = `
.c0 {
  padding: 0;
  list-style: none;
}

.c3 {
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-flex: 30%;
  -ms-flex: 30%;
  flex: 30%;
  text-align: right;
  font-size: 12px;
  line-height: 20px;
}

.c2 {
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-flex: 70%;
  -ms-flex: 70%;
  flex: 70%;
  font-size: 12px;
  line-height: 20px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: left;
}

.c5 {
  height: 4px;
  width: 100%;
  border-radius: 2px;
  background: #F5F5F5;
  position: relative;
}

.c5::after {
  content: '';
  width: NaN%;
  height: 4px;
  border-radius: 2px;
  -webkit-transition: background 0.2s ease;
  transition: background 0.2s ease;
  position: absolute;
  max-width: 100%;
  left: 0;
}

.ciKsuu > .c4::after {
  background: #47DB99;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  cursor: default;
  margin-bottom: 8px;
}

.c6 > .c4::after {
  background: #2C2C2C;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  cursor: default;
  margin-bottom: 8px;
}

.c1 > .c4::after {
  background: #FABADA;
}

<ul
  className="c0"
>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      China
    </p>
    <span
      className="c3"
    >
      359.0M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      United States of America
    </p>
    <span
      className="c3"
    >
      236.5M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      India
    </p>
    <span
      className="c3"
    >
      204.3M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Brazil
    </p>
    <span
      className="c3"
    >
      128.2M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Japan
    </p>
    <span
      className="c3"
    >
      89.7M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Russia
    </p>
    <span
      className="c3"
    >
      80.6M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Mexico
    </p>
    <span
      className="c3"
    >
      67.4M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Indonesia
    </p>
    <span
      className="c3"
    >
      48.4M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Pakistan
    </p>
    <span
      className="c3"
    >
      41.4M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c6"
    color="#FABADA"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Other
    </p>
    <span
      className="c3"
    >
      1.1G
    </span>
    <div
      className="c4 c5"
    />
  </li>
</ul>
`;

exports[`<CategoryWidget /> renders with data 1`] = `
.c0 {
  padding: 0;
  list-style: none;
}

.c3 {
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-flex: 30%;
  -ms-flex: 30%;
  flex: 30%;
  text-align: right;
  font-size: 12px;
  line-height: 20px;
}

.c2 {
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-flex: 70%;
  -ms-flex: 70%;
  flex: 70%;
  font-size: 12px;
  line-height: 20px;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: left;
}

.c5 {
  height: 4px;
  width: 100%;
  border-radius: 2px;
  background: #F5F5F5;
  position: relative;
}

.c5::after {
  content: '';
  width: NaN%;
  height: 4px;
  border-radius: 2px;
  -webkit-transition: background 0.2s ease;
  transition: background 0.2s ease;
  position: absolute;
  max-width: 100%;
  left: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  cursor: default;
  margin-bottom: 8px;
}

.c1 > .c4::after {
  background: #47DB99;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  cursor: default;
  margin-bottom: 8px;
}

.c6 > .c4::after {
  background: #2C2C2C;
}

<ul
  className="c0"
>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      China
    </p>
    <span
      className="c3"
    >
      359.0M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      United States of America
    </p>
    <span
      className="c3"
    >
      236.5M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      India
    </p>
    <span
      className="c3"
    >
      204.3M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Brazil
    </p>
    <span
      className="c3"
    >
      128.2M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Japan
    </p>
    <span
      className="c3"
    >
      89.7M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Russia
    </p>
    <span
      className="c3"
    >
      80.6M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Mexico
    </p>
    <span
      className="c3"
    >
      67.4M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Indonesia
    </p>
    <span
      className="c3"
    >
      48.4M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c1"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Pakistan
    </p>
    <span
      className="c3"
    >
      41.4M
    </span>
    <div
      className="c4 c5"
    />
  </li>
  <li
    className="c6"
    color="#47DB99"
    onClick={[Function]}
    selected={true}
  >
    <p
      className="c2"
    >
      Other
    </p>
    <span
      className="c3"
    >
      1.1G
    </span>
    <div
      className="c4 c5"
    />
  </li>
</ul>
`;

exports[`<CategoryWidget /> renders without crashing 1`] = `
.c0 {
  padding: 0;
  list-style: none;
}

<ul
  className="c0"
/>
`;
